### DISCUSSION TITLE
[@types/alpinejs] Can use interface for `Alpine.data`

### DISCUSSION THREAD

**Discussion by fedemp** [DefinitelyTyped/DefinitelyTyped#66751] [Issues with a @types package] [answered]
This works.

```ts
type MyForm = {
	error: string;
	handleAuth: (err: auth0js.Auth0Error | null) => void;
	handleSubmit: (evt: SubmitEvent) => void;
	isFetching: boolean;
	success: boolean;
}

const myForm = (): MyForm => ({
	handleSubmit: noop,
	handleAuth,
	isFetching: false,
	error: "",
	success: false,
});

Alpine.data("myform", myForm);
```

This won't.

```ts
interface MyForm {
	error: string;
	handleAuth: (err: auth0js.Auth0Error | null) => void;
	handleSubmit: (evt: SubmitEvent) => void;
	isFetching: boolean;
	success: boolean;
}

const myForm = (): MyForm => ({
	handleSubmit: noop,
	handleAuth,
	isFetching: false,
	error: "",
	success: false,
});

Alpine.data("myform", myForm);
```

TS reports that
> Argument of type `() => MyForm` is not assignable to parameter of type `(...args: unknown[]) => AlpineComponent<Record<string | symbol, unknown>>`.
     Type `MyForm` is not assignable to type `AlpineComponent<Record<string | symbol, unknown>>`.
       Type `MyForm` is not assignable to type `Record<string | symbol, unknown>`.
         Index signature for type 'string' is missing in type `MyForm`. [2345]

[a comment by an automated account is not shown]

**Comment by ekwoka** [comment]
Is this the latest version of the types?

You don't need to define a separate type/interface, btw.

**Reply by fedemp** [reply]
Yes. I have just updated types.

I should have attached my tsconfig.json. Here it is.

```json
{
  "extends": "@tsconfig/strictest/tsconfig.json",
  "compilerOptions": {
    "allowJs": false,
    "checkJs": false,
    "noEmit": true,
    "module": "ESNext",
    "lib": ["ESNext"]
  },
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Strictest"
}
```

And it extends https://github.com/tsconfig/bases/blob/main/bases/strictest.json

**Reply by ekwoka** [reply, marked as answer]
I looked into it, and yes, you just can't use interfaces. Because of some of the alpine magic that is needed to make types work (like having persist work, magics available, etc) interfaces just don't work as a direct response.

This is by design in how TypeScript works: https://github.com/microsoft/TypeScript/issues/15300 Old discussion. It hasn't changed.

Currently, the only adjustment possible to the types to make this work is basically just making it so the data function can return ANYTHING.

I'll investigate if there would be a way we could give a base interface your can extend that would give similar protections that we can incorporate that way.

But basically, interfaces aren't meant to be used in the way you're trying to use it here. The appropriate structure is a type.
